File: examples/java/comm_time/Master.java
/*
 * Master of a basic master/slave example in Java
 *
 * Copyright 2006,2007,2010 The SimGrid Team. All rights reserved. 
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. 
 */

import simgrid.msg.*;

public class Master extends simgrid.msg.Process {
   public void main(String[] args) throws JniException, NativeException {
      if (args.length < 4) {
	 Msg.info("Master needs 4 arguments");
	 System.exit(1);
      }
      
      int tasksCount = Integer.valueOf(args[0]).intValue();		
      double taskComputeSize = Double.valueOf(args[1]).doubleValue();		
      double taskCommunicateSize = Double.valueOf(args[2]).doubleValue();
      
      int slavesCount = Integer.valueOf(args[3]).intValue();
      
      Msg.info("Hello! Got "+  slavesCount + " slaves and "+tasksCount+" tasks to process");
      
      for (int i = 0; i < tasksCount; i++) {
	 Task task = new Task("Task_" + i, taskComputeSize, taskCommunicateSize);
	 if (i%1000==0)
	   Msg.info("Sending \"" + task.getName()+ "\" to \"slave_" + i % slavesCount + "\"");
	 task.send("slave_"+(i%slavesCount));
      }
      
      Msg.info("All tasks have been dispatched. Let's tell everybody the computation is over.");
      
      for (int i = 0; i < slavesCount; i++) {
	 FinalizeTask task = new FinalizeTask();
	 task.send("slave_"+(i%slavesCount));
      }
      
      Msg.info("Goodbye now!");
    }
}

File: examples/cxx/basic/Master.cxx
#include "Master.hpp"
#include "BasicTask.hpp"
#include "FinalizeTask.hpp"
#include <Host.hpp>
#include <HostNotFoundException.hpp>

#include <Msg.hpp>

#include <iostream>
using namespace std;


#ifndef BUFFMAX
#define BUFFMAX 260
#endif

MSG_IMPLEMENT_DYNAMIC(Master, Process);

int Master::main(int argc, char** argv)
{
	int channel = 0;
	char buff[BUFFMAX + 1] = {0};
	int numberOfTasks;		
	double taskComputeSize;		
	double taskCommunicateSize;
	StringHelper s;

	if (argc < 3) 
	{
		error("Master needs 3 arguments");
		exit(1);
	}
	
	info("Hello");
	
	int slaveCount = 0;
	Host* slaves = NULL;

	
	info( TEXT_("argc=") + TEXT_(argc));
	
	for (int i = 0; i< argc; i++)	    
		info(TEXT_("argv:") + TEXT_(argv[i]));
	
	sscanf(argv[0],"%d", &numberOfTasks);
	
	sscanf(argv[1],"%lg", &taskComputeSize);
	
	sscanf(argv[2],"%lg", &taskCommunicateSize);
	
	BasicTaskPtr* todo = new BasicTaskPtr[numberOfTasks];
	
	for (int i = 0; i < numberOfTasks; i++) 
		todo[i] = new BasicTask((TEXT_("Task_") + TEXT_(i)), taskComputeSize, taskCommunicateSize);
	
	slaveCount = argc - 3;
	slaves = new Host[slaveCount];
	
	for(int i = 3; i < argc ; i++)  
	{
		try 
		{
			slaves[i-3] = Host::getByName(argv[i]);
		}
		
		catch(HostNotFoundException e) 
		{
			cerr << e.toString() <<". Stopping Now!" << endl;
			exit(1);
		}
	}

	info(TEXT_("Got slave(s) :") + TEXT_(slaveCount));
	
	for (int i = 0; i < slaveCount; i++)
		info(TEXT_("\t") + TEXT_(slaves[i].getName()));
	
	 info(TEXT_("Got ") + TEXT_(numberOfTasks) + TEXT_(" task to process."));
	
	for (int i = 0; i < numberOfTasks; i++) 
	{
		info(TEXT_("Sending \"") + TEXT_(todo[i]->getName()) + TEXT_("\" to \"") + TEXT_(slaves[i % slaveCount].getName()) + TEXT_("\""));
	
		if(!strcmp(Host::currentHost().getName(), slaves[i % slaveCount].getName())) 
			info("Hey ! It's me ! ");

		slaves[i % slaveCount].put(channel, todo[i]);
	}
	
	info("Send completed");
	
	info("All tasks have been dispatched. Let's tell everybody the computation is over.");
	
	for (int i = 0; i < slaveCount; i++) 
			slaves[i].put(channel, new FinalizeTask());

	delete[] todo;
	delete[] slaves;
	
	info("Goodbye now!");

	delete this;

	return 0;
}